### ISSUE TITLE
Add support for ChangeStream.try_next

### ISSUE THREAD

**Issue by ben-dyer** [zookzook/elixir-mongodb-driver#100]
Some drivers support a `try_next` method which sends a single `getMore` command and doesn't block (cf. [pymongo](https://pymongo.readthedocs.io/en/stable/api/pymongo/change_stream.html#pymongo.change_stream.ChangeStream.try_next)). What do you think of adding support for this in the ChangeStream module?

**Comment by ben-dyer** [comment]
For context, one use case is to allow to fetch the next at most n docs in pull-based manner, something like this
```elixir
Enumerable.reduce(
  1..max_calls,
  {:cont, {[], state[:change_stream]}}, # from watch_collection
  fn index, {acc_docs, acc_stream} ->
    {next_batch, new_stream} = ChangeStream.try_next(acc_stream, opts)
    new_docs = acc_docs ++ next_batch
    cursor_empty? = length(next_batch) == 0
    has_max_docs? = length(new_docs) >= max_docs
    case {has_max_docs?, cursor_empty?} do
      {false, false} -> {:cont, {new_docs, new_stream}}
      _ -> {:halt, {new_docs, new_stream}}
    end
  end
)
```
I gave this a shot myself, basically just adding a `try_next` method taking a `%ChangeStream` and opts, calling `get_more` similarly as in `next_fun` from enumerable protocol, and then returning an updated `%ChangeStream`.

I think I could send a PR if you are interested in supporting this

**Comment by zookzook** [comment]
Hey!

thank you for your time. PRs are always welcome. I will be on holiday for the next few weeks. After that I will check the PR.

**Comment by zookzook** [comment]
I checked this issue again and I don't really unterstand it. The ChangeStream is not really blocking, because you can use it in a GenServer send messages, if you receiving new updates, like in the example:

https://github.com/zookzook/elixir-mongodb-driver/blob/master/examples/change_stream/lib/change_stream.ex

You can always spawn a process and call a function that blocks to make it unblocking. Is this working for you?

**Comment by p-mongo** [comment]
Normally the motivation for `try_next` is to provide applications with the ability to persist the updated resume token when a particular getMore doesn't return any changes, such that when a change stream is resumed the resume point hasn't fallen off the oplog. The user side of this is described [here](https://docs.mongodb.com/ruby-driver/master/reference/change-streams/#resuming-a-change-stream).

**Comment by ben-dyer** [comment]
> I checked this issue again and I don't really unterstand it. The ChangeStream is not really blocking, because you can use it in a GenServer and send messages, if you receiving new updates, like in the example:
> 
> https://github.com/zookzook/elixir-mongodb-driver/blob/master/examples/change_stream/lib/change_stream.ex
> 
> You can always spawn a process and call a function that blocks to make it unblocking. Is this working for you?

First of all, apologies for not getting back to this sooner I got caught up on another project, but got back to this recently.

I'm working on a GenStage where the producer pulls from the change stream and consumer writes to an external api. Iterating on the change stream as in the example works perfectly, and running it in a separate task works too - I did this in a first version dumping change stream into an in memory queue.

What I wanted is not to eagerly load from the change stream into application memory (push based model) but only to fetch as much as matches the current demand (pull based model).

My app is just a poc now, so it's lightweight and there is no queue in the middle, so iterating on the change stream can grow the memory arbitrarily.

but `Enum.take` .

With only one consumer per producer the error model is fairly simple

**Comment by ben-dyer** [comment]
> Normally the motivation for try_next is to provide applications with the ability to persist the updated resume token when a particular getMore doesn't return any changes, such that when a change stream is resumed the resume point hasn't fallen off the oplog. The user side of this is described here.

Thank you for sharing that. Here is a sketch of the flow I wanted to build:

- GenStage consumer is ready for 100 more items from change stream, sends demand to producer
- GenStage producer tries to fetch 100 items from the change stream. If the change stream has fewer than 100 documents, say only 10, then it should immediately return the 10 documents to the consumer and as long as there is remaining demand, schedule another request after a certain interval.
- GenStage consumer receives 10 items and sends them to external api
- GenStage consumer persists somewhere the latest successfully posted item's resume token
- When application is restarted it fetches the perisisted resume token; if the token is stale then the application needs to do something else such as snapshot the database or collection, ideally this happens rarely

I couldn't find anything documented that said not to use `try_next` for this kind of demand-based consumption of the change stream, hence I was interested to build a poc of this. I found that the pymongo driver gives the flexibility to do this using `try_next`, but I haven't found any workaround yet for this driver. I did play with `Enum.take`, `Stream.take` and `StreamSplit` as well, but I don't believe they worked for what I wanted.

Besides internal `_try_next` methods, pymongo also publicly exposes `try_next` it as part the api, which suggested to me that a use case like this might also be feasible. `tryNext` is not a part of the official change stream spec, but I think that might be just because it is not universal (appropriate in all programming languages) - they have this to say:

> There is something to be said for an API that allows cooperation by default. The model in which a call to next only blocks until any response is returned (even an empty batch), allows for interruption and cooperation (e.g. interaction with other event loops).
https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst

My main justification for suggesting this feature is to try to match pymongo functionality, to enable this kind of tinkering in Elixir as well as languages with official drivers. In particular, I'm not positive yet that the POC described above would be production-quality usage of Mongo change streams.

**Comment by p-mongo** [comment]
I believe `try_next` in general is a handy feature. If I am remembering correctly we implemented it in the Ruby driver for internal use before it became necessary for resume token support as exposed by 4.2 or 4.4 servers.

I'm not very familiar with blocking/non-blocking iteration in Elixir, so probably whatever opinions I may offer on your particular use case wouldn't have much reasoning behind them.

**Comment by zookzook** [comment]
Don't get me wrong, but I don't think we need the try_next function. The example I mentioned above already fulfills all the requirements. Changes of the documents and the token are reported as messages. I guess, the `try_next` function is more specific to the language environment like ruby or python. Both language are single threaded and are using a Global Interpreter Lock. The Elixir and Erlang the world is very different. You can start a lot of processes and everything runs  run in parallel. This is a very basic feature. The process is a first class citizen  in Erlang. In the example of the try_next usage I see only a polling loop. This is typical for such programming environments. But in Erlang/Elixir you don't need to poll in a such way, you communicate via messages, so the GenServer sends messages and reports changes. Instead of polling by calling a try_next *function* again and again and wasting CPU usage, we just wait for the message so the scheduler is able to distribute the CPU to other processes, that are more active. 

The GenServer in my example can be extended and broadcast the message via `Phoenix.PubSub` and other processes can subscribe to the messages.  You can feed a GenState-Server with that messages to get a back-pressure feature as well.

So it is properly not possible to compare an OOP/single threaded driver like ruby/python with an Erlang/Elixir driver implementation. And unfortunately the specification of the driver uses an OOP style with classes and objects.

**Comment by ben-dyer** [comment]
> I guess, the try_next function is more specific to the language environment like ruby or python. Both language are single threaded and are using a Global Interpreter Lock. 

I see what you mean, there are probably many more common use cases for try_next in a GIL language, but I think the use case I am proposing here is different.

> Don't get me wrong, but I don't think we need the try_next function. The example I mentioned above already fulfills all the requirements.

The main thing that to me is still missing for me here is that if the database is under high load then I think the application memory can grow unboundedly, whereas the GenStage application described above would have a clear maximum memory requirement based on the configuration of GenStage max demand parameters.

(If the consumer code happens inside the Enum.each callback itself then this wouldn't be an issue, but throughput is better in my use case if the consumer processes the documents in batches, so in that case the callback would be dumping documents into a queue somewhere.)

> Instead of polling by calling a try_next function again and again and wasting CPU usage, we just wait for the message so the scheduler is able to distribute the CPU to other processes, that are more active.

Apologies if I misunderstand since I am new with this - I thought iterating on the change stream is more or less doing "while true, call next_fun" (suppose the batch size is 1). Since try_next would basically be the same as next_fun, I was thinking using try_next would be less greedy of cpu resources since it would only be called when the consumer is ready to receive more events and producer has received that demand.

> Don't get me wrong, but I don't think we need the try_next function. The example I mentioned above already fulfills all the requirements. Changes of the documents and the token are reported as messages.

I agree that try_next is not needed, per se. It is possible to get the data from point A to point B without try_next - one can build an application/process which is responsible for capturing the change data and putting in some form of queue (kafka, sqs, phoenix pubsub, whatever) using the existing driver; then a GenStage/Broadway app or process could read from that queue with backpressure. (Afaik Phoenix PubSub is push-based, so this one might be awkward to use with GenStage which processes in batches, but for instance Broadway SQS client is pull-based.) This is what is typically done, I think.

The basis for wanting try_next is to try a simpler architecture which eliminates the queue entirely, and there is backpressure on the process reading from the change stream itself. I would expect some tradeoff between memory/throughput and simplicity/performance compared to using a queue, but this is what I have been interested in trying.

**Comment by zookzook** [comment]
I guess, if you set the batch_size to one then you get what you want, because only one document is return by a getMore command. If you have 100 Documents and you set the batch size to one, what would the different between using try_next and get_more? Each command will return one document and if all are processed, then try_next would return nil and get_more will block. Where is the back pressure feature when using try_next? Both function will work as fast as possible and  may overload the next processing stages. 

So, you can build a GenServer, set the batch size to one, fetch only one document and wait to be asked for the next document. Then you process the next document. I don't see the point, why try_next makes the task easier.

**Comment by ben-dyer** [comment]
> If you have 100 Documents and you set the batch size to one, what would the different between using try_next and get_more? Each command will return one document and if all are processed, then try_next would return nil and get_more will block

As far as I know getMore command won't block and try_next is basically just a single getMore command. I think there is another older concept in the spec called get_more so apologies if my terminology is off, but at any rate I believe `get_more_cmd` from change_stream.ex was not blocking if there are no documents available, based on my experience forking this driver.

> Where is the back pressure feature when using try_next? Both function will work as fast as possible and may overload the next processing stages.

The back pressure is just what's built into GenStage. Basically GenStage comes with some guarantee to never ask the producer for more than a fixed number (`max_demand`) of documents before having handled the previous ones. In order to integrate with that, the producer would need to be able to try to fetch N documents, but be willing to accept fewer than that without waiting for the rest. This is what try_next does in pymongo (when the batch size is N).

If you are interested I can try to make my poc of this public, it is working but still hasn't been tested very extensively at this point.

**Comment by p-mongo** [comment]
Setting batch size to 1 would be quite inefficient if an update contains multiple changes (or many changes).

With the existing APIs, how does an application retrieve the current resume token from the change stream after each getMore if the getMore didn't return any documents? If the API is blocking, and the application called getMore, it would block until getMore returned something, right?

**Comment by zookzook** [comment]
> Setting batch size to 1 would be quite inefficient if an update contains multiple changes (or many changes).
> 
> With the existing APIs, how does an application retrieve the current resume token from the change stream after each getMore if the getMore didn't return any documents? If the API is blocking, and the application called getMore, it would block until getMore returned something, right?

Please take a look to the example. The GenServer implements all features already. The batch size is an option for the back pressure feature. If you need a back pressure then something in your stage is inefficient at all.

**Comment by zookzook** [comment]
> The other possible difference between try_next and getMore is that getMore normally blocks for 5 seconds on the server side if there aren't any changes to be returned, but try_next can return immediately if there aren't any changes that have already been read from the network (provided the driver performs requisite buffering).

As mentioned before: The GenServer example is the key. The server can broadcast messages about the token and the documents and other processes can subscribe to it. The try_next is not needed in the Elixir world. 

Regarding to the title of the this issue: "Add support for ChangeStream.try_next" I would close this issue, because we definitely do not need such a function. 

@ben-dyer: I guess the use case is very specific, so I would suggest that we add your poc to the examples.

**Comment by ben-dyer** [comment]
> @ben-dyer: I guess the use case is very specific, so I would suggest that we add your poc to the examples.

I would be happy to add another example, but IMO the GenServer example is already enough to get started in case of using a queue to interface with GenStage. 

> The other possible difference between try_next and getMore is that getMore normally blocks for 5 seconds on the server side if there aren't any changes to be returned

Not sure, but I don't remember observing that kind of blocking in [get_more](https://github.com/zookzook/elixir-mongodb-driver/blob/master/lib/mongo/change_stream.ex#L127) .

> Regarding to the title of the this issue: "Add support for ChangeStream.try_next" I would close this issue, because we definitely do not need such a function.

No worries, I can live with having a fork for what I want to do. Thank you for reviewing the issue :bowing_man:
